[WIP do not merge] Kube-Proxy Library: Breakout a KPNG-like interface "kube proxy lib" from #2104 (#3649)
kubernetes/enhancements: keps/sig-network/2104-kube-proxy-lib/README.md

marked as `provisional` is a working document and subject to change. You can
denote sections that are under active debate as follows:

```
<<[UNRESOLVED optional short context or usernames ]>>
Stuff that is being argued.
<<[/UNRESOLVED]>>
```

When editing KEPS, aim for tightly-scoped, single-topic PRs to keep discussions
focused. If you disagree with what is already in a document, open a new PR
with suggested changes.

One KEP corresponds to one "feature" or "enhancement" for its whole lifecycle.
You do not need a new KEP to move from beta to GA, for example. If
new details emerge that belong in the KEP, edit the KEP. Once a feature has become
"implemented", major changes should get new KEPs.

The canonical place for the latest set of instructions (and the likely source
of this file) is [here](/keps/NNNN-kep-template/README.md).

**Note:** Any PRs to move a KEP to `implementable`, or significant changes once
it is marked `implementable`, must be approved by each of the KEP approvers.
If none of those approvers are still appropriate, then changes to that list
should be approved by the remaining approvers and/or the owning SIG (or
SIG Architecture for cross-cutting KEPs).
-->

> # KEP-2104: kube-proxy library (KEP-2104, breakout)

Reviewer: Fixing....

> # KEP-2104: kube-proxy library (KEP-2104, breakout)

Reviewer: I guess we'd need a new issue for this right @jayunit100 ?

> # KEP-2104: kube-proxy library (KEP-2104, breakout)

Author: ya, i guess for the last thing we do we can clean that ... lets get consensus on the impl first tho i think b4 we gift wrap that part?

[kubernetes/website]: https://git.k8s.io/website

## Summary

After about a year and a half of testing a new kube-proxy implementation (https://github.com/kubernetes-sigs/kpng/) and
collecting sig-network and community feedback, it became clear that an interface for building new kube proxy's without  an
opinionated implementation is desired by the Kubernetes networking commmunity attempting to build specialied networking
tools.  This KEP distills the goals of such an interface and propose it's lifecycle and official support policy for sig-network.

This distillation of a "KPNG like interface" was originally presented by tim hockins in a sig-network breakout session, informally in https://docs.google.com/presentation/d/1Y-tZ4fFC9L2NvtBeiIXD1MiJ0ieg_zg4Q6SlFmIax8w/edit?hl=en&resourcekey=0-SFhIGTpnJT5fo6ZSzQC57g#slide=id.g16976fedf03_0_221.

## Motivation

There have been several presentations, issues, and projects dedicated to reusing kube proxy logic while extending it to embrace
different backend technologies (i.e. NFT, eBPF, openvswitch, and so on).  This KEP attempts to make a library which will facilitate
this type of work.  

> A general solution to this problem is explored in the KPNG project (https://github.com/kubernetes-sigs/kpng/), which exhibits many properties

Author: ok, removing details, but will still give it a "nod" for historical puproses if thats ok :)

that allow for such goals to be accomplished.  These are enabled by:

- A Generic "*Diff store"* which provides a client side, in memory data model for performant, declarative, generic calculation of differences between the Kubernetes networking state space from one time point to another (i.e. a replacement for the implicit change tracking functionality in k/k/pkg/proxy https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/endpoints.go#L161).
- A *library that simplifies consumption* of the Kubernetes network and its topology easy to consume for an individual node, abstracting away API Semantics (like topology) from underlying network routing rules.
- *Definition of types* which have a minimal amount of boiler plate, can be created without using the Kubernetes API data model (and thus used to extend proxying behaviour to things outside of the Kubernetes pod, service, and endpoint semantics), and which can describe routing of Kubernetes VIPs (services) to endpoints in a generic way that is understandable to people who don't work on Kubernetes on a day-to-day basis.

### Goals

> - Build a vendorable repository "kube-proxy-lib" which can be used to make new kube-proxy's.

Author: service proxies ftw :)

- Exemplify the use of such a repository in a mock "backend" which uses the repository to process and respond to changes in the Kubernetes networking state.
- Define a policy around versioning and releasing of "kube-proxy-lib".
- Create a CI system that runs in test-grid which tests kube-proxy-lib compatibility with the latest Kubernetes releases continuously, and which can be used to inform API changes for regressions that break "kube-proxy-lib".

> - Enable the eventual *replacement* of the k/k/pkg/proxy serviceChangeTracker and endpointsChangeTracker related caching structures inside of in-tree kube proxy with this generic library.

Author: done

### Non-Goals

> - Rewrite or decouple the core , in-tree linux Kubernetes kube-proxy implementations, which are relied on by too many users to be tolerant to major changes.

Reviewer 2: This is incompatible with the goal of replacing ServiceChangeTracker/EndpointsChangeTracker above.

In fact, I think that rebasing the iptables, ipvs, and winkernel implementations on top of this library is a _requirement_, not a non-goal. We don't want to be maintaining two separate implementations of "proxy business logic". (The entire point of this project is to reduce the maintenance burden of people keeping up with "proxy business logic".)

Certainly we need to be careful to not destabilize the existing implementations, but it seems to me that the way to do that is to use the existing shared proxy code (ServiceChangeTracker, etc) as the initial seed of the library, and replace it with new code bit by bit (ensuring that unit, integration, and e2e tests pass all along the way).

Reviewer: > In fact, I think that rebasing the iptables, ipvs, and winkernel implementations on top of this library is a requirement, not a non-goal.

I agree here, however some initial changes like adding shared bits for starting and stopping API server handlers and watchers will also need to be ported i.e something like the [Provided interface](https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/types.go#L29) right?

 IMO the very minimum the first POC/ Seed should still provide external users the ability to start writing new backends quickly not just replace the client side caching mechanisms.

> - Force a new architecture for the standard kube-proxy on to naive users.

Reviewer 2: I assume this means "don't force Kubernetes end users to deploy kube-proxy in a 'split-brain' architecture as in Mikael's original KEP", but that's not at all clear unless you're already aware of that KEP / kpng.

Maybe "Making any incompatible architectural changes to the existing kube-proxy implementations" or something like that?

Reviewer: I'm adding a few here.

## Proposal

We propose to build a kubernetes-sigs/kube-proxy-lib repository.  This repository will be vendored by people wanting to build a new Kube proxy, and provide them with:

> - A vendorable golang library that defines a few interfaces that can be easily implemented as a new kube proxy, which respond to endpoint slices and services changes.

Reviewer 2: ```suggestion
- A vendorable golang library that defines a few interfaces that can be easily implemented by a new service proxy, which respond to EndpointSlice and Service changes.
```

- Documentation on how to build a kube proxy, based on https://github.com/kubernetes-sigs/kpng/blob/master/doc/service-proxy.md and other similar documents.

> - integration test tooling, similar to the KPNG project's, which allows users to locally implement network routing logic in a small golang program which is not

Reviewer: Done, removed that bit

directly connected to the Kubernetes API server, for local, iterative development of Kubernetes network proxy tooling.

### User Stories (Optional)

#### Story 1

As a networking technology startup I want to make my own kube-proxy implementation but don't want to maintain the logic of talking to the APIServer, caching its data, or calculating an abbreviated/proxy-focused representation of the Kubernetes networking state space.  I'd like a wholesale framework I can simply plug my logic into. 

#### Story 2

> As a Kubernetes maintainer, I don't want to have to understand the internals of a networking backend in order to simulate or write core updates to the logic of the kube-proxy locally.

Author: yeah, updated as a svc proxy maint , pushing shortly

### Notes/Constraints/Caveats (Optional)

- sending the full-state could be resource consuming on big clusters, but it should still be O(1) to
  the actual kernel definitions (the complexity of what the node has to handle cannot be reduced

>   without losing functionality or correctness).

Author: deleting

### Risks and Mitigations

No risks, because we arent removing the in tree proxy as part of this repo, but rather, proposing a library for kube proxy extenders
to use optionally.  There may be risks, eventually, if we write another KEP to *replatform* sig-networks k/k/pkg/proxy implementations

> on top of this library, but that would be in a separate KEP. 

Author: Fair point but, in the interest of "building something that is uncontraversial as a first step", How strongly do you feel about this ?  If i made this argument:

- "sig network needs b"
- "b needs a"
- thus were going to make "sig-network/a" as a stepping stone to us being able to build "sig-network/b" 
- by the way "sig-network/b" is somethin lots of people want, so even if we get hit by a bus after making "b", and we never get around to "a", we still have a useful intermediate deliverable.... 

Is that digestible ? or does it sit funny w/ you ?

> on top of this library, but that would be in a separate KEP. 

Reviewer 2: What I'm saying is that I don't think the library can be definitely "uncontroversial" and "a useful intermediate deliverable" unless we _know_ that it is suited to replacing `pkg/proxy`.

I'm not saying we have to deliver the library and the updated iptables proxy in the same release, I'm saying that rather than doing:
1. come up with a plan for the library
2. implement the library
3. come up with a plan to replace `pkg/proxy` with the new library
4. replace `pkg/proxy` with the new library

we need to do
1. come up with a plan for the library
2. come up with a plan to replace `pkg/proxy` with the new library
3. implement the library
4. replace `pkg/proxy` with the new library

> on top of this library, but that would be in a separate KEP. 

Author: gotcha, ok updated as a risk (and updated the goals to reflect)

## Design Details

### Test Plan

##### Unit tests

##### Integration tests

##### e2e tests

### Graduation Criteria

We will version "kube-proxy-lib" as 1.0 once it is known to be powering at least one backend proxy which can be run by an end userr, which is able to pass the Kubernetes sig-network (non disruptive) and Conformance suites, including Layer 7 and serviceType=Loadbalancer tests which currently run in the prow sig-network e2es.

## Production Readiness Review Questionnaire

### Dependencies

This project will depend on the Kubernetes client-go library to acquire Service, Endpoints, EndpointSlices, and other
networking API objects.

### Scalability

The ability to scale this service will be equivalent to that of the existing kube-proxy, insofar as the fact that
it will watch the same endpoints (as the existing kube-proxy) and generally then be used to forward to traffic to a single
backend loadbalancing technology (i.e. ebpf, nft, iptables, ...) as does the existing kube-proxy daemonset.

###### Will enabling / using this feature result in any new API calls?

No

###### Will enabling / using this feature result in introducing new API types?

> Yes. There will be an in-memory API that is supported by this library, that is incremented overr time

Author: ok, added "No" + clarified
